sdk/eventhub/azure-eventhubs/azure/eventhub/aio/_consumer_client_async.py

```diff
@@ -0,0 +1,187 @@
+        :param on_error: The callback function that is called when there is error during receiving events.
+        :param on_partition_initialize: The callback function that is called right before
+         a partition is initialized.
+        :param on_partition_close: The callback function that is called when this receive method finishes receiving
+         from a partition.
+        :rtype: None
+
+        Example:
+            .. literalinclude:: ../samples/test_examples_eventhub_async.py
+                :start-after: [START eventhub_consumer_client_receive_async]
+                :end-before: [END eventhub_consumer_client_receive_async]
+                :language: python
+                :dedent: 4
+                :caption: Receive events from the EventHub.
+        """
+        async with self._lock:
+            error = None
+            if (consumer_group, '-1') in self._event_processors:
+                error = ValueError("This consumer client is already receiving events from all partitions for"
+                                 " consumer group {}. "
+                                 "Shouldn't receive from any other partitions again".format(consumer_group))
+            elif partition_id is None and any(x[0] == consumer_group for x in self._event_processors):
+                error = ValueError("This consumer client is already receiving events for consumer group {}. "
+                                 "Shouldn't receive from all partitions again".format(consumer_group))
+            elif (consumer_group, partition_id) in self._event_processors:
+                error = ValueError("This consumer is already receiving events from partition {} for consumer group {}. "
+                                 "Shouldn't receive from it again.".format(partition_id, consumer_group))
+            if error:
+                log.warning(error)
+                raise error
+
+            event_processor = EventProcessor(
+                self, consumer_group, on_event,
+                partition_id=partition_id,
+                partition_manager=self._partition_manager,
+                error_handler=on_error,
+                partition_initialize_handler=on_partition_initialize,
+                partition_close_handler=on_partition_close,
+                initial_event_position=initial_event_position or EventPosition("-1"),
+                polling_interval=self._load_balancing_interval,
+                owner_level=owner_level,
+                prefetch=prefetch,
+                track_last_enqueued_event_properties=track_last_enqueued_event_properties,
+            )
+            if partition_id:
+                self._event_processors[(consumer_group, partition_id)] = event_processor
+            else:
+                self._event_processors[(consumer_group, "-1")] = event_processor
+        try:
+            await event_processor.start()
+        finally:
+            await event_processor.stop()
```

Comment:
It seems that start will run indefinitely - does that mean we only hit the finally block if an error occurs?

Reply:
Mainly for some interruptions. For instance, KeyboardInterupt.
Another example is using task.cancel( ). This cancel will raise CancelledError so the finally block is executed.
```python
task = asyncio.create_task(client.receive(...))
asyncio.sleep(10)
task.cancel( )
```